dvc/repo/ls.py

```diff
@@ -1,79 +1,127 @@
 import os
+import stat
 
 from dvc.exceptions import PathMissingError, OutputNotFoundError
 
 
 @staticmethod
-def ls(url, target=None, rev=None, recursive=None, outs_only=False):
+def ls(
+    url, target=None, rev=None, recursive=None, outs_only=False,
+):
+    """Methods for getting files and outputs for the repo.
+
+    Args:
+        url (str): The repo url
+        target (str, optional)
+        rev (str, optional)
+        recursive (bool, optional)
+        outs_only (bool, optional): Shows only DVC-artifacts
+
+    Returns:
+        list of `entry`
+
+    Notes:
+        `entry` is a dictionary with structure
+        {
+            "path": str,
+            "isout": bool,
+            "isdir": bool,
+            "isexec": bool,
+        }
+    """
     from dvc.external_repo import external_repo
     from dvc.repo import Repo
     from dvc.utils import relpath
 
     with external_repo(url, rev) as repo:
         target_path_info = _get_target_path_info(repo, target)
-        result = []
+        fs_nodes = []
         if isinstance(repo, Repo):
-            result.extend(_ls_outs_repo(repo, target_path_info, recursive))
+            fs_nodes.extend(_ls_outs_repo(repo, target_path_info, recursive))
 
         if not outs_only:
-            result.extend(_ls_files_repo(target_path_info, recursive))
+            fs_nodes.extend(_ls_files_repo(target_path_info, recursive))
 
-        if target and not result:
+        if target and not fs_nodes:
             raise PathMissingError(target, repo, output_only=outs_only)
 
-        def prettify(path_info):
-            if path_info == target_path_info:
-                return path_info.name
-            return relpath(path_info, target_path_info)
+        fs_nodes = {n["path_info"]: n for n in fs_nodes}.values()
 
-        result = list(set(map(prettify, result)))
-        result.sort()
-    return result
+        def get_entry(fs_node):
+            path_info = fs_node["path_info"]
+            path = (
+                path_info.name
+                if path_info == target_path_info
+                else relpath(path_info, target_path_info)
+            )
+            return {
+                "path": path,
+                "isout": fs_node.get("isout", False),
+                "isdir": fs_node.get("isdir", False),
+                "isexec": fs_node.get("isexec", False),
+            }
 
+        entries = [get_entry(fs_node) for fs_node in fs_nodes]
+        entries.sort(key=lambda f: f["path"])
```

Comment:
I use in-place sort so it is fast enough (even faster I assume), also it looks more readable - every line contains one instruction

Reply:
I benched it, `sorted()` is faster. The rest is more about taste, I prefer to avoid mutable ops, so I might be biased )

Reply:
Just resolve it if you wish.

Reply:
I've done [a small research](https://gist.github.com/JIoJIaJIu/19c954140a4cd9c6908f096c76ca495a) with **10^6** shuffled list
```bash
$ python 1.py 
         1000004 function calls in 0.822 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.822    0.822 1.py:11(comp_and_sort)
        1    0.284    0.284    0.413    0.413 1.py:12(<listcomp>)
  1000000    0.128    0.000    0.128    0.000 1.py:8(get_entry)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.409    0.409    0.409    0.409 {method 'sort' of 'list' objects}
```
```bash
$ python 1.py 2
         1000003 function calls in 0.820 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.820    0.820 1.py:17(map_and_sorted)
  1000000    0.128    0.000    0.128    0.000 1.py:8(get_entry)
        1    0.692    0.692    0.820    0.820 {built-in method builtins.sorted}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

They are the same (if we construct the list), what is logical cause in my approach I create list first(`<listcomp>`) and then sort it(`sort`), and in yours it's done in one place (`builtins.sorted`) 

About immutability - it makes sense and I think it would be better to use `sorted` then, thanks :+1: